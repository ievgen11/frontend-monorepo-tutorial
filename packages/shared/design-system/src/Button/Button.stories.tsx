import * as React from "react";
import { withKnobs, text, boolean } from "@storybook/addon-knobs";
import { action } from "@storybook/addon-actions";

import { Button } from "./index";

export default {
  title: "Button",
  component: Button,
  decorators: [withKnobs],
};

export const Default = () => (
  <Button
    title={text("Title", "Track Your Package")}
    disabled={boolean("Disabled", false)}
    onClick={action("onClick")}
  >
    {text("Text", "Track Your Package")}
  </Button>
);
