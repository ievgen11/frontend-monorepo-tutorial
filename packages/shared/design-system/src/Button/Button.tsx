import React from "react";
import styled from "styled-components";

const StyledButton = styled.button`
  background: #f0f0f0;
  border: none;
  color: #212529;
  padding: 12px 24px;
  border-radius: 32px;
  font-weight: 500;
  font-size: 12px;
  white-space: nowrap;
  text-overflow: ellipsis;
  box-shadow: -4px -4px 18px 6px #fff, 0 6px 12px 0 rgba(0, 0, 0, 0.2),
    6px 0 12px 0 rgba(0, 0, 0, 0.2), -6px 0 12px 0 rgba(0, 0, 0, 0.2);
  transition: all 0.1s ease-in-out;
  cursor: pointer;

  @media only screen and (min-width: 768px) {
    padding: 18px 32px;
    font-size: 14px;
  }

  &:hover {
    background-color: #fff;
    transform: translate(0, -1px);
  }

  &:focus,
  &:active {
    background-color: #fff;
    box-shadow: -4px -4px 12px 8px #f0f0f0, 0 2px 6px 0 rgba(0, 0, 0, 0.2),
      2px 0 6px 0 rgba(0, 0, 0, 0.2), -2px 0 6px 0 rgba(0, 0, 0, 0.2);
    transform: translate(0, 1px);
  }

  &:focus {
    outline-style: none;
  }

  &:disabled {
    cursor: not-allowed;
    background-color: #c8c8c8;
    color: #a0a0a0;
  }
`;

export interface ButtonProps {
  children?: React.ReactNode;
  onClick?: () => void;
  className?: string;
  disabled?: boolean;
  title?: string;
}

export const Button: React.FC<ButtonProps> = ({
  children,
  onClick = (): null => null,
  ...props
}): JSX.Element => (
  <StyledButton onClick={(): void => onClick()} {...props}>
    {children}
  </StyledButton>
);
