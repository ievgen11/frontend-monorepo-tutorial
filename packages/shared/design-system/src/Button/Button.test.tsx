import * as React from "react";
import * as renderer from "react-test-renderer";
import { render, fireEvent } from "@testing-library/react";
import { Button } from "./Button";

const ButtonTestID = "ButtonTestID";

describe("<Button />", () => {
  test("WOW! Button displays text!", async () => {
    const { getByTestId } = render(
      <Button data-testid={ButtonTestID}>This is a test!</Button>
    );
    const button = await getByTestId(ButtonTestID);

    await fireEvent.click(button);

    expect(button).toHaveTextContent("This is a test!");
  });
  it("WHOA! Snapshot is nice!", () => {
    const tree = renderer.create(<Button />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
