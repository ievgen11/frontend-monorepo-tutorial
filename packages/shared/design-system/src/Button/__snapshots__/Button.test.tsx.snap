// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> WHOA! Snapshot is nice! 1`] = `
<button
  className="sc-bdfBwQ hXIWBA"
  onClick={[Function]}
/>
`;
